File: algorithms/243.shortest_word_distance/shortest_word_distance.cc
<<<<<<< HEAD
class Solution {
public:
    int shortestDistance(vector<string>& words, string word1, string word2) {
        int idx1 = -1, idx2 = -1, res = words.size();
        for (int i = 0; i < words.size(); ++i) {
            if (words[i] == word1) {
                idx1 = i;
                if (idx2 != -1) res = min(res, idx1 - idx2); 
            } else if (words[i] == word2) {
                idx2 = i;
                if (idx1 != -1) res = min(res, idx2 - idx1);
            }
        }
        return res;
    }
};
// Shortest Word Distance
class Solution {
public:
  int shortestDistance(vector<string>& words, string word1, string word2) {
    int n = words.size(), p = -n, q = -n, r = n;
    for (int i = 0; i < n; i++)
      if (words[i] == word1)
        p = i, r = min(r, p-q);
      else if (words[i] == word2)
        q = i, r = min(r, q-p);
    return r;
  }
=======
class Solution {
public:
    int shortestDistance(vector<string>& words, string word1, string word2) {
        int idx1 = -1, idx2 = -1, res = words.size();
        for (int i = 0; i < words.size(); ++i) {
            if (words[i] == word1) {
                idx1 = i;
                if (idx2 != -1) res = min(res, idx1 - idx2); 
            } else if (words[i] == word2) {
                idx2 = i;
                if (idx1 != -1) res = min(res, idx2 - idx1);
            }
        }
        return res;
    }
};
// Shortest Word Distance
class Solution {
public:
  int shortestDistance(vector<string>& words, string word1, string word2) {
    int n = words.size(), p = -n, q = -n, r = n;
    for (int i = 0; i < n; i++)
      if (words[i] == word1)
        p = i, r = min(r, p-q);
      else if (words[i] == word2)
        q = i, r = min(r, q-p);
    return r;
  }
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74
};

File: algorithms/243.shortest_word_distance/shortest_word_distance.java
public class Solution {
    public int shortestDistance(String[] words, String word1, String word2) {
        int i=-1, j=-1;
        int min = Integer.MAX_VALUE;
        for(int k=0; k<words.length; k++) {
            if(words[k].equals(word1) ) {
                if(j!=-1 && i==-1) {
                    min = k - j;
                } else {
                    min = Math.min(min, k-j);
                }
                i = k;
            } else if(words[k].equals(word2)){
                if(i!=-1 && j==-1) {
                    min = k - i;
                } else {
                    min = Math.min(min, k-i);
                }
                j = k;
            }
        }        
        return min;
    }
}
public class Solution {
    public int shortestDistance(String[] words, String word1, String word2) {
        int idx1 = -1, idx2 = -1, distance = Integer.MAX_VALUE;
        for(int i = 0; i < words.length; i++){
            if(words[i].equals(word1)){
                idx1 = i;
                // 第一次写入idx就先不比较
                if(idx2 != -1) distance = Math.min(distance, idx1 - idx2);
            }
            if(words[i].equals(word2)){
                idx2 = i;
                // 第一次写入idx就先不比较
                if(idx1 != -1) distance = Math.min(distance, idx2 - idx1);
            }
        }
        return distance;
    }
}